### PR TITLE
kvutils: Remove references to Daml-LF Archive from kvutils protos [KVL-1167]

### PR DESCRIPTION
**Pull request by hubert-da**
### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:245]: Given that the hash is a small part of the archive, I suggest a custom parser here (or in `ArchiveParser`).

Does `ArchiveParser` actually exist?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala:183]: Please update this comment to explain we still parse the archive every time.

We may want to introduce a cache for parsing archives.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Raw.scala
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Raw.scala]: I gave up on this type, anyway.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitterSpec.scala
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitterSpec.scala:7]: Nope, restored.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitterSpec.scala
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitterSpec.scala]: This is just a test, added tho.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala]: Discussion about caching to be continued here: https://github.com/digital-asset/daml/pull/12168#discussion_r770496675

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:245]: This is actually the approach we discussed with @remyhaemmerle-da. WDYT? Also, this will go to the `kv-support` soon, I believe.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala:212]: I think this is a valuable remark. Based on my discussion with @cocreature and @remyhaemmerle-da, parsing should be cheap. I'm not sure if we want to optimize for that before even knowing if the impact is significant.

**Review by cocreature — APPROVED** [review]
Looks sensible, I’ll leave it up to you what benchmarks you want to run before merging.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala
Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:245]: @remyhaemmerle-da We can make a custom parser based on the following protobuf. Whether it’s significantly faster I dunno but at least it’s plausible.
```
message Archive {
  string hash = 4;
}
```

Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:245]: please put optimization in kv-support. the language-team may move those latter in the archive module.

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala:212]: It's not just an optimization for the computer, it also de-duplicates code, which is an optimization for my brain when reading it. :wink:

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:245]: > @remyhaemmerle-da We can make a custom parser based on the following protobuf. Whether it’s significantly faster I dunno but at least it’s plausible.
I am completely fine for any optimization, as long one have sufficient confidence it is actually faster. 

If we are so concern about performance, I will also skip the string conversion. 
```
message Archive {
  bytes hash = 4;
}
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala:184]: We always call `decodeArchive`, but usually it'll be cached. The parser does not cache.

Perhaps the comment can be updated to clarify this.

Comment (cocreature) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala:184]: `decodearchive` will not be cached. The engine caches so that you will not call `lookupPackage` and you skip both decoding & parsing.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:245]: > I would expect that decoding a field to a `ByteString` is pretty cheap so I’m not sure the performance impact of this is particularly large.

Depends how you got there. `CodedInputStream#readBytes` might just take a substring, or it might copy, depending on the underlying data structure.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitterSpec.scala
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitterSpec.scala:180]:
```suggestion
    val hashes = 
      archives
        .iterator()
        .asScala
        .map(archive => Conversions.extractHashFromArchive(Raw.Archive(archive)))
        .toSet
    hashes shouldBe committedPackages.toSet[String]
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala:184]: Then was the original comment wrong?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala:212]: Ok, let me reshuffle this a bit 😉

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/CommitterModelConformanceValidator.scala:184]: No, it's my addition that slightly doesn't make sense. I will make it clear that the `lookupPackage` is called once per archive.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:245]: Here the underlying is a `ByteString`, isn't it ? 
So I expect it take a substring. 
If you can show me that in you usecase, it is copying I am fine for writing an optimization.

Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:245]: I am good with holding off on this until we actually need it (if ever).

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala:236]: I did what I could. There are still 2 functions that parse the archives. These 2 functions will most likely go to `kv-support`. I don't want to store the archives to avoid Daml-LF references.

**Review by SamirTalwar-DA — APPROVED** [review]

**Comment by hubert-da** [comment]
> I’ll leave it up to you what benchmarks you want to run before merging.

As agreed with Samir, we decided not to benchmark it before merging, as it's just about package uploads that don't happen very often and it would require additional effort to propagate changes to the integration kit.

**Pull request merged**
